Nested data structure from bo file

katiebuntic opened the pull request:
Having a start at using dataclasses for creating a nested structure from the BO file

stefpiatek commented on dataclass_function.py:
```diff
@@ -0,0 +1,86 @@
+
+
+@dataclass
+class Main:
```
maybe `School` would be clearer?

katiebuntic replied:
Oh yes agreed!

katiebuntic replied:
Great, thanks! I will have a further look at nested pandas structure now

katiebuntic commented on dataclass_function.py:
```diff
@@ -0,0 +1,86 @@
+    month6_bl_result: int
+    month6_wv_result: int
+    month6_bas_result: int
+    RRED_QC_Parameters: str
```
Need to change this to lowercase

stefpiatek reviewed (commented):
This looks good! I think we can either use this directly or I wonder if we can use https://pypi.org/project/pandas-dataclasses/ to have a nested pandas structure. Worth having a quick look to see if installing that and we can use pandas

stefpiatek commented:
👌 Might actually be better to have 3 separate data frames that you can join together, the Teacher data type would have `RREDUserID` as the index, but also have the `school_id` as a column. Then the `School` data type would have `school_id` as an index?

katiebuntic commented:
> 

Sounds good!

katiebuntic commented:
@stefpiatek just updated it to the pandas way - is this what you had in mind? I will do some tests with it with the BO file data

stefpiatek commented on dataclass_function.py:
```diff
@@ -0,0 +1,94 @@
+class Pupil(AsFrame):
+    """Pupil information"""
+
+    pupilno: Index[complex]
```
oh what's complex?

stefpiatek reviewed (commented):
🎉 Yeah perfect. Great, hopefully should be easy to convert from the massive data frame into our nested structure

stefpiatek reviewed (commented):
Looking good, will wait for approval until we have functions and tests that can convert the input file to the data structure.

If we wanted to be super sure as well, we could have a data type of the original input file, so we're explicit about what we're expecting.

stefpiatek commented on dataclass_function.py:
```diff
@@ -1,41 +1,43 @@
-    entry_dob: Data[datetime.datetime]
+    pupil_no: Index[str]
+    rred_user_id: Data[str]
+    assessi_engtest2: Data[float]
```
out of interest does this work from pandas? Would be preferable if it did
```suggestion
    assessi_engtest2: Data[Int32Dtype]
```


 If not then should add a comment the first time so that we know why we're using floats

katiebuntic replied:
This doesn't work :( so will just add a comment on why

stefpiatek commented on dataclass_function.py:
```diff
@@ -1,41 +1,43 @@
+    assessiii_engtest11: Data[float]
+    assessiii_engtest9: Data[float]
+    assessiii_iretest4: Data[float]
+    entry_dob: Data[str]
```
Worth checking in review that we haven't lost any datetimes

stefpiatek commented on dataclass_function.py:
```diff
@@ -92,3 +94,108 @@ class School(AsFrame):
+        rrcp_area
+        rrcp_country
+    """
+    school_df = School.new(clmnlist(6), clmnlist(5), clmnlist(4), clmnlist(3))
```
would we need to make the schools unique first?

it may handle it but I imaginged making a new df which is just the school columns, unique and then create all of the schools like this?

You can then remove the school columns for the next step

katiebuntic replied:
Ah yes that would be much nicer, will update this

stefpiatek commented on dataclass_function.py:
```diff
@@ -92,3 +94,108 @@ class School(AsFrame):
+    """
+    I feel like below could be a for loop but not quite sure how
+    """
+    pupil_df = Pupil.new(
```
then once you got to here, I expect you could do something like passing in all of the rows to the Pupil constructor which harry should be able to help with if it gets really hairy

katiebuntic commented:
Have added another file for the Main df `dataclass_function_main.py` which is indexing the school + teacher, then the pupil information is the data. It looks alright but may not be what is needed..

I could not seem to work nesting the three dfs within pandas dataclass (`dataclass_function.py`) - may need some help with this next week if poss @harryjmoss :) 
![image](https://user-images.githubusercontent.com/96536608/222736475-24975c07-6fe8-4a31-b650-35103babc567.png)

harryjmoss commented:
@katiebuntic have gone through this and things are looking good generally! I'd stared at the problem for a while thinking about the nested dataframe structure and came up with a few changes in [this pull request](https://github.com/UCL-ARC/rred-reports/pull/21) - feel free to take as many or as few of these are you like. Will also need some tests here as Stef mentioned.

@stefpiatek apols if this wasn't the nested structure we're ultimately going for!

katiebuntic commented:
@harryjmoss ready for review now :)

stefpiatek commented on src/rred_reports/dataclass_function.py:
```diff
@@ -0,0 +1,189 @@
+        return list(full_data.iloc[:, i])
+
+    all_schools_df = School(clmnlist(6), clmnlist(5), clmnlist(4), clmnlist(3))
+    all_schools_df = all_schools_df.drop_duplicates()  # pylint: disable=E1101
```
Instead of having pylint disabling maybe we should edit our config for pylint. Looks like we can do this [for pandas](https://stackoverflow.com/questions/33961756/disabling-pylint-no-member-e1101-error-for-specific-libraries#35106735)  I think we should be able to do this in the `pyproject.toml` file, under its  `[tool.pylint]` section

stefpiatek commented on src/rred_reports/dataclass_function.py:
```diff
@@ -0,0 +1,189 @@
+
+
+def main():
+    """Entrypoint for generating nested dataframe"""
+    file_path = Path(__file__).resolve().parents[2] / "input" / "templates" / "example_dataset.xlsx"
+    full_nested_df = createdf(file_path)
+
+    unravelled_data = parse_nested_dataframe(full_nested_df)
+
+    school_teacher_df = unravelled_data["school_teacher"]
+    teacher_pupil_df = unravelled_data["teacher_pupil"]
+
+    print_nested_dataframe_contents(school_teacher_df)
+    print_nested_dataframe_contents(teacher_pupil_df)
+
+    return full_nested_df
```
this would be great to have moved to a test, and add the example_dataset. Maybe adding some asserts at the end (doesn't need to be extensive) to make sure that the dimensions the dataframes are right and that a couple of the values are correct.

stefpiatek reviewed (approved):
Looks good, thanks for getting this in 🎉 

I think worth addressing all of the changes in here, happy to chat through if anything is unclear.

katiebuntic commented:
Hi @harryjmoss and @stefpiatek - I have made a couple of changes for review:
- Logger import added in
- Changed all `floats` to `pd.Int32Dtype` in the pandas dataclasses
- Added in the .new when creating the dataframe
- Tests at the end to check all works

There were still some issues with the pylint so I also added in `# pylint: disable=E1121' when creating the dfs - I added in a line to the `pyproject.toml` for this but may need a check as didn't seem to help?

Also, the `example_dataset.xlxs` didn't work when it was in the `test` folder so I have kept it in the main for now

Thank you! :)

stefpiatek commented:
yeah annoying that the pylint config didn't work but hey ho, I'll create an issue for it. 
Cool, I think merge it in 👏 , I might have some time today between things to set up some quick tests for it and get the file working from the test dir
